src/theory/quantifiers/sygus/sygus_random_enumerator.h

```diff
@@ -0,0 +1,140 @@
+ * generated terms approximate a geometric distribution with an expected size of
+ * 1 / p, where p is a parameter specified by the user (defaults to 0.5).
+ */
+class SygusRandomEnumerator : public EnumValGenerator
+{
+ public:
+  /** Constructor. Initializes the enumerator.
+   *
+   * @param tds pointer to term database sygus.
+   * @param tn the sygus type to enumerate.
+   */
+  SygusRandomEnumerator(TermDbSygus* tds, TypeNode tn);
+
+  /** Initialize this class with enumerator `e`. */
+  void initialize(Node e) override;
+
+  /** Inform this generator that abstract value `v` was enumerated. */
+  void addValue(Node v) override { d_cache.insert(v); }
+
+  /**
+   * Get next the next random (T-rewriter-unique) value.
+   *
+   * @return true if a new value is found and loop otherwise.
+   */
+  bool increment() override;
+
+  /** @return the current term. */
+  Node getCurrent() override { return d_currTerm; }
+
+ private:
+  /** Initialize helper utility. */
+  void initializeH();
+
+  /** Generates a random sygus term.
+   *
+   * S ::= 0 | x | (+ S S)
+   *
+   * Assuming we are provided the above grammar, we generate random terms by
+   * starting with a skolem, `z0`, of the given sygus datatype type (grammar
+   * non-terminal). We then flip a coin to determine whether or not we should
+   * replace the skolem with a random constructor that takes arguments (`(+ S
+   * S)` above). For example, if the first 2 coin flips are heads, then we will
+   * replace `z0` with `(+ z1 z2)` and either `z1` or `z2` (randomly chosen)
+   * with `(+ z3 z4)`. So, we will end up with either `(+ (+ z3 z4) z2)` or `(+
+   * z1 (+ z3 z4))`. We keep doing so until we get a tails. At which point, we
+   * replace all remaining skolems with random no-argument constructors (`0` and
+   * `x` above) and return the resulting sygus term. So, if we get a tails at
+   * the third flip in the previous example, then `incrementH` may return `(+ 1
+   * (+ x x))`, `(+ (+ 1 1) 1)`, or any of the other 14 possible terms.
+   *
+   * \note If a skolem's datatype type does not have a no-argument constructor,
+   * it is replaced with a ground value using `mkGroundValue` utility.
+   * \note If a skolem's datatype type does not have a constructor that takes an
+   * argument (e.g., `S ::= 0 | 1 | x | (+ 2 y)`), it is replaced with a random
+   * no-argument constructor. So `incrementH` may return a term before getting a
+   * tails.
+   *
+   * @return a randomly generated sygus term.
+   */
+  Node incrementH();
+
+  /**
+   * Returns smallest (in terms of size) term equivalent (up to rewriting) to
+   * the given sygus term.
+   */
+  Node getMin(Node n);
+
+  /** Pointer to term database sygus. */
+  TermDbSygus* d_tds;
+  /** The type to enumerate. */
+  TypeNode d_tn;
+  /** Random number generator. */
+  std::mt19937_64 d_gen;
+  /** Bernoulli distribution. */
+  std::bernoulli_distribution d_bernoulli;
```

Comment:
Not sure if this and the utility below this are available in the standard utilites.  Maybe @aniemetz can comment.